### PR TITLE
Create RestoreSummaryRequest objects in parallel

### LINKED ISSUE #8737: Restore requests should be created in parallel
## Details about Problem

When restoring a large number of projects, the [DependencyGraphSpecRequestProvider](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs) [creates requests](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs#L62) in a sequential `foreach` loop.  However, the request creation can be slow depending on the complexity of a project.  The creation of these requests should be done in parallel to spread the load across the machine's CPU cores.

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): All

NuGet version (x.x.x.xxx): Any

dotnet.exe --version (if appropriate):  Any

VS version (if appropriate): Any

OS version (i.e. win10 v1607 (14393.321)): Any

Worked before? If so, with which NuGet version:

## Detailed repro steps so we can see the same problem

1. Profile the creation of a large number of restore requests

The CPU is under utilized since the creation of the request objects is serial.

### Verbose Logs

Please include verbose logs (NuGet.exe <COMMAND> -verbosity detailed | dotnet.exe <COMMAND> --verbosity diag | etc...)

### Sample Project

Very helpful if you can zip a project and paste into this issue!

### PR DESCRIPTION
**Pull request by jeffkl**
When restoring a lot of projects, the creation of RestoreSummaryRequest objects is slow.  By creating them in parallel large restores are 15% faster.

I also had to update the return value of IPreLoadedRestoreRequestProvider.CreateRequests to IReadOnlyCollection since I'm using a ConcurrentBag<T> in the Parallel.ForEach which doesn't support IList<T>

## Bug

Fixes: https://github.com/NuGet/Home/issues/8737
Regression: No
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: I've implemented the usage of a `Parallel.ForEach` loop so that the creation of requests is done on all CPUs.

## Testing/Validation

Tests Added: No  
Reason for not adding tests:  Existing tests should cover this scenario
Validation:  Profiling shows the creation of requests happening on multiple threads now which reduces restore times by up to 20%

### CODE REVIEW

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs [hunk 1]
```diff
@@ -56,10 +57,18 @@ private IReadOnlyList<RestoreSummaryRequest> GetRequestsFromItems(RestoreArgs re
...
-            foreach (var projectNameToRestore in dgFile.Restore)
+            var parallelOptions = new ParallelOptions
+            {
+                // By default, max degree of parallelism is -1 which means no upper bound.
+                // Limiting to processor count reduces task context switching which is better
+                MaxDegreeOfParallelism = Environment.ProcessorCount
+            };
+
+            // Parallel.Foreach has an optimization for Arrays, so calling .ToArray() is better and adds almost no overhead
+            Parallel.ForEach(dgFile.Restore.ToArray(), parallelOptions, projectNameToRestore =>
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs:71]: I have a concern with adding additional scheduled work. 
We blast the threadpool in too many places. 

How does this perform in smaller solutions? Any meaningful improvement there?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs:71]: This change also "clashes" with https://github.com/NuGet/NuGet.Client/pull/3086. 
Once that's merged, I'd expect the gains from this to be lower.

Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs:71]: This happens before requests are scheduled.  In my profiling, everything happens in parallel except this part of the restore.  The perf gain is much more apparent in large restores, smaller restores will probably not appear to be any faster.

Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs:71]: In my testing, I disabled the loading of settings per project and still see about .5 to 1 second improvement for 700 projects.  So I think this is still needed along with #3086

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs:71]: > This happens before requests are scheduled

That's a good point. :)

**Review by dtivel — APPROVED** [review]

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs [hunk 2]
```diff
@@ -33,14 +34,14 @@ public class DependencyGraphSpecRequestProvider : IPreLoadedRestoreRequestProvid
             _providerCache = providerCache;
         }
 
-        public Task<IReadOnlyList<RestoreSummaryRequest>> CreateRequests(RestoreArgs restoreContext)
+        public Task<IReadOnlyCollection<RestoreSummaryRequest>> CreateRequests(RestoreArgs restoreContext)
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs]: This is a breaking change on a public method on a public class. I find it very, very difficult to imagine why anyone would call this API, but what's the performance impact if you keep the same datatype and copy the results into a list after the `Parallel.ForEach` completes?

Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs]: Thanks for the concern.  I did more digging and found that `ToList<T>()` would be slow but `ToArray()` is fast and gets rid of the need to change the return type.  I've updated the PR!

**Review by zivkan — APPROVED** [review]

**Review by nkolev92 — APPROVED** [review]

**Pull request merged**